internal/annotation/filter.go

```diff
@@ -0,0 +1,210 @@
+	return newFilter(key, opts)
+}
+
+// NewFalsyEventHandler returns an event handler that enqueues objects
+// that do not have annotation with key string key or whose value is falsy.
+func NewFalsyEventHandler(key string, opts Options) (handler.EventHandler, error) {
+	opts.truthy = false
+	return newEventHandler(key, opts)
+}
+
+// NewTruthyPredicate returns a predicate that passes objects
+// that do have annotation with key string key and whose value is truthy.
+func NewTruthyPredicate(key string, opts Options) (predicate.Predicate, error) {
+	opts.truthy = true
+	return newFilter(key, opts)
+}
+
+// NewTruthyEventHandler returns an event handler that enqueues objects
+// that do have annotation with key string key and whose value is truthy.
+func NewTruthyEventHandler(key string, opts Options) (handler.EventHandler, error) {
+	opts.truthy = true
+	return newEventHandler(key, opts)
+}
+
+func defaultOptions(opts *Options) {
+	if opts.Log == nil {
+		opts.Log = logf.Log
+	}
+}
+
+// newEventHandler returns a filter for use as an event handler.
+func newEventHandler(key string, opts Options) (handler.EventHandler, error) {
+	f, err := newFilter(key, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	f.hdlr = &handler.EnqueueRequestForObject{}
+	return handler.Funcs{
+		CreateFunc: func(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+			if f.Create(evt) {
+				f.hdlr.Create(evt, q)
+			}
+		},
+		UpdateFunc: func(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+			if f.Update(evt) {
+				f.hdlr.Update(evt, q)
+			}
+		},
+		DeleteFunc: func(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+			if f.Delete(evt) {
+				f.hdlr.Delete(evt, q)
+			}
+		},
+		GenericFunc: func(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+			if f.Generic(evt) {
+				f.hdlr.Generic(evt, q)
+			}
+		},
+	}, nil
+}
+
+// newFilter returns a filter for use as a predicate.
+func newFilter(key string, opts Options) (*filter, error) {
+	defaultOptions(&opts)
+
+	// Make sure the annotation key and eventual value are valid together.
+	if err := validateAnnotation(key, opts.truthy); err != nil {
+		return nil, err
+	}
+
+	f := filter{}
+	f.key = key
+	// Falsy filters return true in all cases except when the annotation is present and true.
+	// Truthy filters only return true when the annotation is present and true.
+	f.ret = !opts.truthy
+	f.log = opts.Log.WithName("pause")
+	return &f, nil
+}
+
+func validateAnnotation(key string, truthy bool) error {
+	fldPath := field.NewPath("metadata", "annotations")
+	annotation := map[string]string{key: fmt.Sprintf("%v", truthy)}
+	return validation.ValidateAnnotations(annotation, fldPath).ToAggregate()
+}
+
+// filter implements a filter for objects with a truthy "paused" annotation (see Key).
+// When this annotation is removed or value does not evaluate to "true",
+// the controller will see events from these objects again.
+type filter struct {
+	key  string
+	ret  bool
```

Comment:
Non-blocking thought:

It might be generally useful to add a `Not(in predicate.Predicate) predicate.Predicate` wrapper in C-R that inverts the logic of the input predicate.

If that existed, would you need the `ret` field?

I alluded to this in the C-R PR where I added the `Or` and `And` wrappers [here](https://github.com/kubernetes-sigs/controller-runtime/pull/1043#issuecomment-656872379), but I never added it because there hasn't been a use case for it.

Reply:
I don't think you'd need `ret` here in that case. Given that operator-lib is pre v1.0, are you ok with adding this as-is now then removing if `Not` is added to c-r?